src/ai/defaultai.cc

```diff
@@ -5856,7 +5830,22 @@ BuildingNecessity DefaultAI::check_building_necessity(BuildingObserver& bo,
 			} else {
 				bo.cnt_target = 1 + static_cast<int32_t>(mines_.size() + productionsites.size()) / 30;
 			}
+			// Determine whether we need more buildings due to low stocklevel
+			const uint32_t current_stocklevel = (get_stocklevel(bo, gametime));
+			const uint8_t stocklevel_threshhold =
+			   std::abs(management_data.get_military_number_at(179));
 
+			if (current_stocklevel > stocklevel_threshhold &&
+			    persistent_data->remaining_basic_buildings.count(bo.id) == 0) {
+				bo.cnt_target -= 1;
+			}
+
+			if (current_stocklevel < stocklevel_threshhold / 2) {
+				bo.cnt_target += 1;  // increase target
+				// here we increase primary priority if stock gets low to get higher prio on each field
```

Comment:
Do you not afraid that cnt_target will go to the sky?
what about check like
if bo.total_count() == bo.cnt_target
Also check for unoccupied sites, if any we should not increase cnt_target

Reply:
it can't go to the sky as it is always reinitialised in line 5831 or line 5829.     
However checking unoccupied sites is a valid point will incorporate this

Reply:
Indeed, but then all you can do here is increase by 1 in addition to what is set above as you mentioned. My idea was to kind of remember cnt_target between runs, so what about to modify 5831:

`bo.cnt_target = std::max(bo.cnt_target, 1 + static_cast<int32_t>(mines_.size() + productionsites.size()) / 30);`

Reply:
Or incorporate hard upper limit e.g. : 5 + static_cast<int32_t>(mines_.size() + productionsites.size()

Reply:
as we are running this check much faster as the building could be built and start working I think remeber the target between runs really might be dangerous to go to the max and we would build too much of them.  I'd rather include the size of trainingsites here as a measure for the need of food then pushing the target too much.

Reply:
You would not increase the target until:

- no construction sites
- no unoccupied
- total_count() = cnt_target


So it seems safe, but as you think